packages/js/components/src/image-gallery/image-gallery.tsx

```diff
@@ -1,33 +1,164 @@
+import { ImageGalleryToolbar } from './index';
+import { ImageGalleryChild, MediaUploadComponentType } from './types';
 
 export type ImageGalleryProps = {
-	children: JSX.Element | JSX.Element[];
+	children: ImageGalleryChild | ImageGalleryChild[];
 	columns?: number;
+	onRemove?: ( props: {
+		removeIndex: number;
+		removedItem: ImageGalleryChild;
+	} ) => void;
+	onReplace?: ( props: {
+		replaceIndex: number;
+		previousItem: ImageGalleryChild;
+		newItem: ImageGalleryChild;
+	} ) => void;
+	onOrderChange?: ( items: ImageGalleryChild[] ) => void;
+	MediaUploadComponent?: MediaUploadComponentType;
 } & React.HTMLAttributes< HTMLDivElement >;
 
 export const ImageGallery: React.FC< ImageGalleryProps > = ( {
 	children,
 	columns = 4,
+	onOrderChange = () => null,
+	onRemove = () => null,
+	onReplace = () => null,
+	MediaUploadComponent = MediaUpload,
 }: ImageGalleryProps ) => {
+	const [ toolBarItem, setToolBarItem ] = useState< string | null >( null );
+	const [ orderedChildren, setOrderedChildren ] = useState<
+		ImageGalleryChild[]
+	>( [] );
+
+	useEffect( () => {
+		if ( ! children ) {
+			return;
+		}
+		setOrderedChildren(
+			( Array.isArray( children ) ? children : [ children ] ).map(
+				( child, index ) =>
+					cloneElement( child, { key: child.key || String( index ) } )
+			)
+		);
+	}, [ children ] );
+
+	const updateOrderedChildren = ( items: ImageGalleryChild[] ) => {
+		setOrderedChildren( items );
+		onOrderChange( orderedChildren );
+	};
+
+	const moveItem = useCallback(
+		( fromIndex: number, toIndex: number ) => {
+			updateOrderedChildren(
+				moveIndex< ImageGalleryChild >(
+					fromIndex,
+					toIndex,
+					orderedChildren
+				)
+			);
+		},
+		[ orderedChildren ]
+	);
+
+	const removeItem = useCallback(
+		( removeIndex: number ) => {
+			updateOrderedChildren(
+				orderedChildren.filter( ( _, index ) => index !== removeIndex )
+			);
+			onRemove( {
+				removeIndex,
+				removedItem: orderedChildren[ removeIndex ],
+			} );
+		},
+		[ orderedChildren ]
+	);
+
+	const replaceItem = useCallback(
+		( replaceIndex: number, newSrc: string, newAlt: string ) => {
+			const newChildren = [ ...orderedChildren ];
+			newChildren.splice(
+				replaceIndex,
+				1,
+				cloneElement( orderedChildren[ replaceIndex ], {
+					src: newSrc,
+					alt: newAlt,
+				} )
+			);
+			onReplace( {
+				replaceIndex,
+				previousItem: orderedChildren[ replaceIndex ],
+				newItem: newChildren[ replaceIndex ],
+			} );
+			updateOrderedChildren( newChildren );
+		},
+		[ orderedChildren ]
+	);
+
 	return (
 		<div
 			className="woocommerce-image-gallery"
 			style={ {
 				gridTemplateColumns: 'min-content '.repeat( columns ),
 			} }
 		>
-			<Sortable isHorizontal>
-				{ Children.map( children, ( child ) => (
-					<SortableHandle>{ child }</SortableHandle>
-				) ) }
+			<Sortable
+				isHorizontal
+				onOrderChange={ ( items ) => {
+					updateOrderedChildren( items );
+				} }
+				notSortableIndexes={ [ 0 ] }
```

Comment:
I think a compositional API might be slightly advantageous over props here.  Could we make a component `NonSortableItem` or something to that effect that we can wrap items with?

Reply:
Sounds good @joshuatf , this was possible by utilizing classes again (so we don't mandate a particular prop in child components) and moving it into the `ImageGalleryItem` component. 

Done in https://github.com/woocommerce/woocommerce/pull/34871/commits/46e5c7b1e1ce90586e62ab75fa6392a97c854a28.